core/pta/device.c

```diff
@@ -88,14 +88,21 @@ static TEE_Result invoke_command(void *pSessionContext __unused,
 {
 	TEE_Result res = TEE_SUCCESS;
 	uint32_t rflags = 0;
+	/*
+	 * FIXME: this should also be true if CFG_RPMB_ANNOUNCE_PROBE_CAP is
```

Comment:
I don't think this can be fixed, at least not without adding more exchanges between the kernel driver and OP-TEE. 
How big a problem is this?

Reply:
Could Linux set an `OPTEE_{SMC|FFA}_NSEC_CAP_RPMB_RPC` capability flag in `OPTEE_*_EXCHANGE_CAPABILITIES` command, when RPMB driver is supported?

This could ease addressing cases where Linux supports (and needs) RPMB subsystem while U-Boot does not (and rely on its tee-supplicant to be used.

Reply:
> Could Linux set an `OPTEE_{SMC|FFA}_NSEC_CAP_RPMB_RPC` capability flag in `OPTEE_*_EXCHANGE_CAPABILITIES` command, when RPMB driver is supported?
> 

Unfortunately not, due to:
``` C
        if (args->a1 & ~OPTEE_SMC_NSEC_CAP_UNIPROCESSOR) {
                /* Unknown capability. */
                args->a0 = OPTEE_SMC_RETURN_ENOTAVAIL;
                return;
        }
```

> This could ease addressing cases where Linux supports (and needs) RPMB subsystem while U-Boot does not (and rely on its tee-supplicant to be used.

tee-supplicant is always ready in U-Boot so it could do `optee_enumerate_devices(PTA_CMD_GET_DEVICES_RPMB)`, followed by `optee_enumerate_devices(PTA_CMD_GET_DEVICES_SUPP)` without lying.

Reply:
> > This could ease addressing cases where Linux supports (and needs) RPMB subsystem while U-Boot does not (and rely on its tee-supplicant to be used.
> 
> tee-supplicant is always ready in U-Boot so it could do optee_enumerate_devices(PTA_CMD_GET_DEVICES_RPMB), followed by optee_enumerate_devices(PTA_CMD_GET_DEVICES_SUPP) without lying.

It does not work well since eMMC device driver may not be initialized when optee driver is initialized. `PTA_CMD_GET_DEVICES_RPMB` makes OP-TEE core to init the RPMB driver that ends in device not being found (when eMMC is not inited), so in the end U-Boot `PTA_CMD_GET_DEVICES_RPMB` fails. This enumeration needs to wait both eMMC and optee device drivers are initialized.

Reply:
Regarding:
```
 if (args->a1 & ~OPTEE_SMC_NSEC_CAP_UNIPROCESSOR) {
                /* Unknown capability. */
                args->a0 = OPTEE_SMC_RETURN_ENOTAVAIL;
                return;
        }
```
there could some evolution here to allow NSEC to pass other capability bits. New kernel that set other bits and get a `OPTEE_SMC_RETURN_ENOTAVAIL` respond would know passing extra NSEC caps feature is not supported by OP-TEE and would need to re-issue a new EXCHANGE_CAPS command without these bits set.
Of course, there could also be new dedicated EXCHANGE_NSEC_CAPS smc/ffa command, or some new commands in PTA_DEVICE to exchange dedicated info between both sides.

Reply:
> It does not work well since eMMC device driver may not be initialized when optee driver is initialized. `PTA_CMD_GET_DEVICES_RPMB` makes OP-TEE core to init the RPMB driver that ends in device not being found (when eMMC is not inited), so in the end U-Boot `PTA_CMD_GET_DEVICES_RPMB` fails. This enumeration needs to wait both eMMC and optee device drivers are initialized.

Will it? I thought U-Boot initialized devices on demand.

Reply:
> Regarding:
> 
> ```
>  if (args->a1 & ~OPTEE_SMC_NSEC_CAP_UNIPROCESSOR) {
>                 /* Unknown capability. */
>                 args->a0 = OPTEE_SMC_RETURN_ENOTAVAIL;
>                 return;
>         }
> ```
> 
> there could some evolution here to allow NSEC to pass other capability bits. New kernel that set other bits and get a `OPTEE_SMC_RETURN_ENOTAVAIL` respond would know passing extra NSEC caps feature is not supported by OP-TEE and would need to re-issue a new EXCHANGE_CAPS command without these bits set. Of course, there could also be new dedicated EXCHANGE_NSEC_CAPS smc/ffa command, or some new commands in PTA_DEVICE to exchange dedicated info between both sides.

The usual way to do it is like for `OPTEE_SMC_FUNCID_ENABLE_ASYNC_NOTIF`.

Reply:
> > This enumeration needs to wait both eMMC and optee device drivers are initialized.
>
> Will it? I thought U-Boot initialized devices on demand.

Testing few configurations I see this constraint actually comes from system services we embedded in OP-TEE. Having SCMI services in OP-TEE make U-Boot to probe the optee driver very early, possibly before mmc driver is initialized. In our stm32mp configurations where the eMMC clocks/regulators depend on OP-TEE SCMI service, the dependency is even stronger.